### PR TITLE
Add Nushell

### PR DESCRIPTION
**Pull request by hustcer**
## Checklist:

- [x] **I am adding a new language.**
  - [x] The extension of the new language is used in hundreds of repositories on GitHub.com.
    - Search results for each extension:
      - https://github.com/search?q=path%3A*.nu&type=code
     
  - [x] I have included a real-world usage sample for all extensions added in this PR:
    - Sample source(s):
      - https://github.com/nushell/nushell/blob/main/.github/workflows/release-pkg.nu
      - https://github.com/nushell/nu_scripts/blob/main/make_release/this_week_in_nu_weekly.nu
    - Sample license(s): [MIT](https://github.com/nushell/nushell/blob/main/LICENSE)
  - [x] I have included a syntax highlighting grammar: [Nushell Grammar](https://github.com/hustcer/nu-grammar/blob/main/grammars/tmLanguage.json)
      
  - [x] I have added a color
    - Hex value: `#4E9906`
    - Rationale: It's the main theme color of Nushell logo from: https://github.com/nushell
    
  - [x] I have updated the heuristics to distinguish my language from others using the same extension.

### CODE REVIEW

**Comment by lildude** [comment]
How is Nushell different from Nu which Linguist already supports and has done so for many many years? From a cursory naïve look, it appears to be the same thing with Nushell often being referred to as Nu, even in the [Nushell book](https://www.nushell.sh/book/).

**Comment by hustcer** [comment]
> How is Nushell different from Nu which Linguist already supports and has done so for many many years? From a cursory naïve look, it appears to be the same thing with Nushell often being referred to as Nu, even in the [Nushell book](https://www.nushell.sh/book/).

They are totally two different language with the same name, and they are created by two different team. That's why I updated `lib/linguist/heuristics.yml` to distinguish Nushell from Nu using the same extension.

File: test/test_heuristics.rb [hunk 1]
```diff
@@ -705,6 +705,12 @@ def test_nl_by_heuristics
     })
   end
 
+  def test_nu_by_heuristics
+    assert_heuristics({
+      "Nushell" => all_fixtures("Nushell", "*.nu")
+    })
+  end
```
Comment (lildude) [inline test/test_heuristics.rb:712]: This needs to also confirm the heuristics fallback to Nu for the Nu files.

File: lib/linguist/languages.yml [hunk 2]
```diff
@@ -4524,6 +4524,22 @@ Nunjucks:
   tm_scope: text.html.nunjucks
   ace_mode: nunjucks
   language_id: 461856962
+Nushell:
+  type: programming
+  color: "#5EA602"
+  extensions:
+  - ".nu"
+  - ".nush"
```
Comment (lildude) [inline lib/linguist/languages.yml:4532]: There are no `.nush` files currently indexed so we can't include it:

```suggestion
```

File: lib/linguist/heuristics.yml [hunk 3]
```diff
@@ -478,6 +478,10 @@ disambiguations:
   - language: NL
     pattern: '^(b|g)[0-9]+ '
   - language: NewLisp
+- extensions: ['.nu']
+  rules:
+  - language: Nushell
+    pattern: '^\s*(import|export|module|def|let|let-env) '
```
Comment (lildude) [inline lib/linguist/heuristics.yml:484]: We know the only other match is going to be Nu so we can shortcut things and save the classifier:

```suggestion
    pattern: '^\s*(import|export|module|def|let|let-env) '
  - language: Nu
```

**Review by lildude — CHANGES_REQUESTED** [review]
Thanks for the clarification that this is a completely different language.

We need a few changes though:

1. Please update the template in the OP to state the licenses of the two samples and add a reason for why you have chosen the colour you have.
2. You need to add the cached license file for the grammar. This should have been created when you run `script/add-grammar` (you can manually run the command [here](https://github.com/github-linguist/linguist/blob/5a0c74277548122267d84283910abd5e0b89380e/script/add-grammar#L134) if you can't find the file).
3. See inline comments.

**Comment by hustcer** [comment]
> 
> 1. Please update the template in the OP to state the licenses of the two samples and add a reason for why you have chosen the colour you have.

I don't know what's **OP**. I have just updated the description of license for the two samples. And updated the color to `#4E9906`, that's the main theme color of nushell logo from: https://github.com/nushell

> 2. You need to add the cached license file for the grammar. This should have been created when you run `script/add-grammar` (you can manually run the command [here](https://github.com/github-linguist/linguist/blob/5a0c74277548122267d84283910abd5e0b89380e/script/add-grammar#L134) if you can't find the file).

Added

> 3. See inline comments.

I have removed `.nush` extension (It's planed to be used in the future, we can add it after that)
For the other two comments, does that mean the Nushell scripts be recognized as `Nu` language? We may want to recognized as `Nushell` to make it different from [Nu](https://github.com/programming-nu/nu)

**Comment by lildude** [comment]
I'm showing my age: OP == original post 😁

> For the other two comments, does that mean the Nushell scripts be recognized as `Nu` language? We may want to recognized as `Nushell` to make it different from [Nu](https://github.com/programming-nu/nu)

No. We need to ensure that your heuristics do not incorrectly identify Nu files as Nushell.

The heurstics are read top-down. Which means we'll try the Nushell heuristic first and if it doesn't match, we move onto the next regex for that extension. If there isn't one, or there isn't a match, we move onto the classifier. The classifier should be considered the last resort as it's trained on the samples and as such may not be accurate.

By adding `- language: Nu` to the end of the heuristics, you tell Linguist that if the `.nu` file it is analysing doesn't match the Nushell regex, it will be identified as Nu and Linguist will then not try analysing the file with the classifier.

The `assert_heuristics` test you've added needs to be updated to test and confirm this behaviour for the Nu samples we already have. You can see how this is done with other tests in this file.

**Comment by hustcer** [comment]
@lildude Thanks for the detailed explanation, I have just updated the PR, Please check it again

**Comment by hustcer** [comment]
Should fix https://github.com/github-linguist/linguist/issues/6214

**Comment by hustcer** [comment]
@lildude Can this PR be **Approved**?

**Review by lildude — APPROVED** [review]
LGTM. Thanks.

Note: this PR will not be merged until close to when the next release is made. See [here](https://github.com/github/linguist/blob/master/docs/troubleshooting.md#my-linguist-pr-has-been-merged-but-github-doesnt-reflect-my-changes) for more details.

**Review by lildude — APPROVED** [review]

**Pull request merged**